File: 1768-mergeStringsAlternately/mergeAlternately.ts
function mergeAlternately(word1: string, word2: string): string {
    let result = [];
    let i = 0;
    let j = 0;
    while (i < word1.length && j < word2.length) {
        result.push(word1[i]);
        i++;
        result.push(word2[j]);
        j++;
    }
    while (i < word1.length) {
        result.push(word1[i]);
        i++;
    }
    while (j < word2.length) {
        result.push(word2[j]);
        j++;
    }
    return result.join("");
};

// Some test cases
console.log(mergeAlternately("abc", "pqr")); // "apbqcr"
console.log(mergeAlternately("ab", "pqrs")); // "apbqrs"
console.log(mergeAlternately("abcd", "pq")); // "apbqcd"

File: 1768-mergeStringsAlternately/merge_alternately.rs
fn merge_alternately(word1: String, word2: String) -> String {
    word1
        .chars()
        .zip(word2.chars())
        .flat_map(|(c1, c2)| [c1, c2])
        .chain(word1.chars().skip(word2.len()))
        .chain(word2.chars().skip(word1.len()))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_merge_alternately() {
        assert_eq!(
            merge_alternately("abc".to_string(), "pqr".to_string()),
            "apbqcr".to_string()
        );
        assert_eq!(
            merge_alternately("ab".to_string(), "pqrs".to_string()),
            "apbqrs".to_string()
        );
        assert_eq!(
            merge_alternately("abcd".to_string(), "pq".to_string()),
            "apbqcd".to_string()
        );
    }
}
